src/main/scala/com/exasol/cloudetl/filesystem/FileSystemManager.scala

```diff
@@ -1,19 +1,61 @@
 package com.exasol.cloudetl.filesystem
 
+import java.io.FileNotFoundException
+import java.nio.file.Path
+
+import scala.collection.mutable.ListBuffer
+
+import org.apache.hadoop.fs.{Path => HadoopPath}
+import org.apache.hadoop.fs.FileStatus
 import org.apache.hadoop.fs.FileSystem
-import org.apache.hadoop.fs.Path
 
-object FileSystemManager {
+/**
+ * A filesystem helper class.
+ */
+final case class FileSystemManager(fileSystem: FileSystem) {
 
-  def globWithLocal(path: java.nio.file.Path, fs: FileSystem): Seq[Path] =
-    globWithPattern(path.toAbsolutePath.toUri.getRawPath, fs)
+  /**
+   * Returns list of files from path with or without globbing.
+   *
+   * It also adds paths from the sub-directories to the returned list.
+   *
+   * @param path input path
+   * @return list of paths
+   */
+  def getFiles(path: String): Seq[HadoopPath] = globStatus(path)
 
-  def globWithPattern(pattern: String, fs: FileSystem): Seq[Path] =
-    glob(new Path(pattern), fs)
+  /**
+   * Returns list of files from local path.
+   */
+  def getLocalFiles(path: Path): Seq[HadoopPath] =
+    getFiles(path.toAbsolutePath().toUri().getRawPath())
+
+  private[this] def globStatus(path: String): Seq[HadoopPath] = {
+    val hadoopPath = new HadoopPath(path)
+    val statuses = fileSystem.globStatus(hadoopPath, HiddenFilesFilter)
+    if (statuses == null) {
+      throw new FileNotFoundException(
+        s"Provided file path '$path' does not exist. Please use valid path."
+      )
+    } else {
+      listStatus(statuses)
+    }
+  }
 
-  def glob(path: Path, fs: FileSystem): Seq[Path] = {
-    val opt = Option(fs.globStatus(path, HiddenFilesFilter))
-    opt.fold(Seq.empty[Path])(_.toSeq.map(_.getPath))
+  @SuppressWarnings(Array("org.wartremover.warts.Recursion"))
+  private[this] def listStatus(paths: Array[FileStatus]): Seq[HadoopPath] = {
+    val files = new ListBuffer[HadoopPath]
+    def addFilesRecursively(fileStatuses: Array[FileStatus]): Unit =
+      fileStatuses.foreach {
+        case fileStatus =>
+          if (fileStatus.isDirectory()) {
+            addFilesRecursively(fileSystem.listStatus(fileStatus.getPath(), HiddenFilesFilter))
+          } else {
+            files += fileStatus.getPath()
+          }
```

Comment:
I'm not sure if it's a good idea to use recursion here. Does scalar do optimizations for that? For very deep nested file structures it can lead to large stacks. 
On the other hand, it's very readable...

Reply:
Something else: You're loading all file names in memory. So for very very large collections of very small files, this could make you run out of memory.

Reply:
Yes, this is also something new. Now thinking, I also do not like it subdirectory listing. 

Users already have mechanism to provide that using globbing if they want to.